std/crypto/chacha20.zig

```diff
@@ -0,0 +1,338 @@
+// Based on public domain Supercop by Daniel J. Bernstein
+
+const mem = @import("../mem.zig");
+const endian = @import("../endian.zig");
+const builtin = @import("builtin");
+
+const QuarterRound = struct {
+    a: usize,
+    b: usize,
+    c: usize,
+    d: usize,
+};
+
+fn Rp(a: usize, b: usize, c: usize, d: usize) QuarterRound {
+    return QuarterRound{
+        .a = a,
+        .b = b,
+        .c = c,
+        .d = d,
+    };
+}
+
+fn rotate(a: u32, b: u5) u32 {
+    return ((a << b) |
+            (a >> @intCast(u5, (32 - @intCast(u6, b))))
+           );
+}
+
+fn salsa20_wordtobyte(input: [16]u32) [64]u8 {
+    var x: [16]u32 = undefined;
+    var out: [64]u8 = undefined;
+
+    for (x) |_, i|
+        x[i] = input[i];
+    const rounds = comptime []QuarterRound{
+        Rp( 0, 4, 8,12),
+        Rp( 1, 5, 9,13),
+        Rp( 2, 6,10,14),
+        Rp( 3, 7,11,15),
+        Rp( 0, 5,10,15),
+        Rp( 1, 6,11,12),
+        Rp( 2, 7, 8,13),
+        Rp( 3, 4, 9,14),
+    };
+    comptime var j: usize = 20;
+    inline while (j > 0) : (j -=2) {
+        for (rounds) |r| {
+            x[r.a] +%= x[r.b]; x[r.d] = rotate(x[r.d] ^ x[r.a], 16);
+            x[r.c] +%= x[r.d]; x[r.b] = rotate(x[r.b] ^ x[r.c], 12);
+            x[r.a] +%= x[r.b]; x[r.d] = rotate(x[r.d] ^ x[r.a],  8);
+            x[r.c] +%= x[r.d]; x[r.b] = rotate(x[r.b] ^ x[r.c],  7);
+        }
+    }
+    for (x) |_, i|
+        x[i] +%= input[i];
+    for (x) |_, i|
+        mem.writeInt(out[4 * i .. 4 * i + 4], x[i], builtin.Endian.Little);
+    return out;
+}
+
+pub fn chaCha20(in: []const u8, key: [32]u8, nonce: [8]u8, out: *[]u8) void {
```

Comment:
that resulted in a segfault in the compiler!: #1372

Reply:
You need to remove the current `out` dereferences in your code such as `out.*`. A slice in zig is just a pointer and a length pair. Since this is already a pointer to memory passing it as a pointer is necessary indirection and not required.